add missing status code of swarm cluster in api 1.24

allencloud opened the pull request:
In the api docs, I found that some status codes daemon side returns are also very important.
These codes have already been implemented by the daemon, while are missing in the API docs.

Since 1.12 is born, these status codes are already implemented. Now 1.12 has been frozen, api 1.24 doc still has  these missing status codes which needs added.

Although status code 406 may be some kind of improper which needs more discussion, I think 1.13 still bears 406 one of the missing status code. Afterwards, we will go further to discuss more about the more accurate one.

**- What I did**
1. Add missing status codes in the API docs 1.24 in the aspect of swam cluster.

Signed-off-by: allencloud <allen.sun@daocloud.io>

allencloud committed:
add missing status code in api docs

Signed-off-by: allencloud <allen.sun@daocloud.io>

ehazlett commented:
LGTM (not a maintainer)

allencloud commented:
ping @thaJeztah @stevvooe @aaronlehmann 

PTAL

If this PR makes sense and gets merged luckily, I think we need to start to make clear that if some status codes are proper which is addressed in https://github.com/docker/docker/pull/26576

thaJeztah commented:
@allencloud I'd like to have @stevvooe's input on this, per https://github.com/docker/docker/pull/26576#issuecomment-260516126

thaJeztah commented:
Oh, looks like this needs a rebase as well

allencloud commented:
Updated @thaJeztah 
PTAL

allencloud commented:
Do I have to update the doc api 1.26? 
@thaJeztah @ehazlett @stevvooe

stevvooe commented:
LGTM (@thaJeztah PTAL)

Let's handle that in #26576.

I think we have to leave v1.24 with 406 codes (I think I misunderstood this before).

allencloud commented:
ping @thaJeztah @vdemeester PTAL

stevvooe commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -6105,6 +6127,12 @@ Remove the secret `id` from the secret store
+
+-   **204** – no Content
+-   **404** – unknown secret
+-   **406** – node is not part of a swarm
```
Looks like one was missed.

stevvooe commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -4558,8 +4558,9 @@ List nodes
-- **200** – no error
-- **500** – server error
+-   **200** – no error
+-   **406** – node is not part of a swarm
```
Should be 503.

allencloud commented:
Thanks for your feedback, @stevvooe . I ever thought that I should fill up all the missing 406 here in doc 1.25 within this PR, then change all of them into 503.

Since it seems to be better to finish 503 here in api doc 1.25, it is OK for me.

Updated this PR for all 406s to be 503s in doc 1.25. PTAL

vdemeester reviewed (approved):
LGTM :tiger:

thaJeztah commented:
@allencloud I don't think the change from 406 to 503 was versioned, so it's only returned in API 1.24 on a 1.12 daemon, but not if the daemon is running 1.13. Should we change them all to 503?

allencloud commented:
@thaJeztah
There is no change in API 1.24, and this PR just added some missing status code of 406 in the api 1.24 docs. Therefore, there is nothing to be versioned in this part.

In addition, there is change which is from 406 to 503 in API 1.25, since PR https://github.com/docker/docker/pull/28532 done by @stevvooe . In this Part, there are two things to be done: 1. update the API 1.25 doc; 2. make this change versioned(this is the part your comment cares above, right? If so, I just updated this part 😄 ).

thaJeztah commented:
> There is no change in API 1.24, and this PR just added some missing status code of 406 in the api 1.24 docs. Therefore, there is nothing to be versioned in this part.

I think there is; if I run a 1.13 daemon, it will return a 406, even on API version 1.24

> make this change versioned

I'm a bit in doubt here; strictly, yes, we cannot change API 1.24. On the other hand, following the discussion, if seems to have been a bug that it returned a 406 in the first place, so I'm not sure which direction should be taken

@stevvooe @bfirsh wdyt?

bfirsh commented:
For anyone else joining in here, @thaJeztah clarified what was going on here in an offline discussion:

* in docker version 1.12.x, API 1.24 it returns 406
* In docker version 1.13.x, API 1.24 + API 1.25 it returns 503

The status code change should absolutely be versioned – that'll break code. v1.25 on any Docker version should return 503, v1.24 on any Docker version should return 406, even if it semantically incorrect. If users want correctness, they can upgrade to v1.25. ;)

allencloud commented:
Hi, @thaJeztah @bfirsh @stevvooe @vdemeester @ehazlett 
Sorry for my delay.

> I think there is; if I run a 1.13 daemon, it will return a 406, even on API version 1.24
I was wrong before, @thaJeztah 's opinion is correct.

Now I update the PR, and this PR involves threes parts:
1. add missing status codes for api 1.24
2. update swagger to add 503
3. version the api change in v1.25

LK4D4 commented:
@allencloud need rebase

allencloud commented:
Rebased, Thanks @LK4D4 
I was wondering if I changed swagger.yml, do I have to execute some command to auto generated some go files. My machine is OSX, and I found that there is `make swagger-gen` in Makefile which uses `docker run`.

thaJeztah commented:
Thanks @allencloud, docs look good to me, but I don't think the _code_ is in place to return a 406 for API 1.24, and a 503 for API 1.25 (both running on the same version of the daemon) https://github.com/docker/docker/blob/master/api/server/httputils/errors.go#L65-L71, correct?

thaJeztah commented:
Maybe I'm confused now, so please confirm if I'm right :D

allencloud commented:
Oh, yeah. This part is some kind of confusing. @thaJeztah 

My understanding and my test told me that:

1. for dockerd 1.13.0+, its provided api v1.24 and api v1.25, and both v1.24 and v1.25 return status code of 503 rather than 406.

2. this part is really different.  for dockerd 1.12.x, its provided api v1.24 just returns status code of 406.

So same api version 1.24, but two different status codes returns from different version of dockerd.

However, I still insist that we should make this change for the long term consideration. But inconsistency we should document some places.

WDYT? @ehazlett @stevvooe @bfirsh @LK4D4

thaJeztah commented:
@allencloud unfortunately, API v1.24 is in use, and documented, so basically we cannot change it. So even on a 1.13 daemon, it should return 406's for API 1.24, even if it's "wrong" to use 406. See https://github.com/docker/docker/pull/28349#issuecomment-262236124

allencloud commented:
> make this change

Maybe we need to change more code to clarify this to make this change. keep current 406 for v1.24, and 503 for v1.25+.

How to implement the separated 406 and 503, I think we had better to do another PR to finish this.

Can we first add the missing status code of 406 in api doc v1.24? Since the original purpose of this PR is to add these codes in docs.

thaJeztah commented:
> Can we first add the missing status code of 406 in api doc v1.24? Since the original purpose of this PR is to add these codes in docs.

Yes, agreed, let's merge this first

thaJeztah reviewed (approved):
LGTM
